src/ir/from_ast/astconv.rs

```diff
@@ -860,94 +850,79 @@ impl<'ctx, 'prog> BuildCtx<'ctx, 'prog> {
             )
             .into(),
         );
+
         for (idx, len) in ports {
-            let idx = idx.expr(self.comp);
-            let start = idx.gte(self.comp.num(0), self.comp);
-            let end = idx.lt(len, self.comp);
-            let in_range = start.and(end, self.comp);
+            let idx = idx.expr(&mut self.comp);
+            let start = idx.gte(self.comp.num(0), &mut self.comp);
+            let end = idx.lt(len, &mut self.comp);
+            let in_range = start.and(end, &mut self.comp);
             cmds.extend(self.comp.assume(in_range, reason))
         }
         cmds
     }
-
-    fn external(ctx: &ir::Context, sig: ast::Signature) -> ir::Component {
-        let mut ir_comp = ir::Component::new(true);
-        ir_comp.src_info = Some(ir::InterfaceSrc::new(sig.name.copy()));
-        let binding = SigMap::default();
-        let mut builder = BuildCtx::new(ctx, &mut ir_comp, &binding);
-
-        // First we declare all the ports
-        let mut cmds = builder.sig(sig);
-        cmds.extend(builder.port_assumptions());
-        ir_comp.cmds = cmds;
-        ir_comp
-    }
-
-    fn comp(
-        ctx: &ir::Context,
-        ir_comp: &mut ir::Component,
-        comp: ast::Component,
-        sigs: &'prog SigMap,
-    ) {
-        let mut builder = BuildCtx::new(ctx, ir_comp, sigs);
-
-        let mut cmds = builder.sig(comp.sig);
-        let body_cmds = builder.commands(comp.body);
-        cmds.extend(builder.port_assumptions());
-        cmds.extend(body_cmds);
-        ir_comp.cmds = cmds;
-    }
 }
 
 pub fn transform(ns: ast::Namespace) -> ir::Context {
-    let mut sig_map = SigMap::default();
-    let main_idx = ns.main_idx();
-    let (mut ns, order) = crate::utils::Traversal::from(ns).take();
-
-    // Walk over signatures and build a SigMap
-    for (idx, sig) in ns.signatures().map(|(_, sig)| sig).enumerate() {
-        sig_map.insert(sig.name.copy(), Sig::from((sig, idx)));
-    }
+    // creates an empty context with the main index.
+    let mut ctx = ir::Context {
+        entrypoint: ns
+            .main_idx()
+            // index main component after all externals
+            .map(|idx| crate::utils::Idx::new(ns.externals().count() + idx)),
```

Comment:
Don't use `crate::` namespace inline. Instead, add `use crate::utils`. 

Also, why is this manually creating `Idx` (which is generally a dangerous thing to do). Is it because we need to do a `checked_add` afterwards?

Reply:
Yeah the checked add requires it, and it also is required for foreigns to properly work as well, because all the idxs need to be generated before they are actually added to a context.